File: src/compiler/backend/targets/pio/PIOPeephole.cpp
/*
 * -----------------------------------------------------------------------------
 * PyMCU Compiler (pymcuc)
 * Copyright (C) 2026 Ivan Montiel Cardona and the PyMCU Project Authors
 *
 * SPDX-License-Identifier: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * -----------------------------------------------------------------------------
 * SAFETY WARNING / HIGH RISK ACTIVITIES:
 * THE SOFTWARE IS NOT DESIGNED, MANUFACTURED, OR INTENDED FOR USE IN HAZARDOUS
 * ENVIRONMENTS REQUIRING FAIL-SAFE PERFORMANCE, SUCH AS IN THE OPERATION OF
 * NUCLEAR FACILITIES, AIRCRAFT NAVIGATION OR COMMUNICATION SYSTEMS, AIR
 * TRAFFIC CONTROL, DIRECT LIFE SUPPORT MACHINES, OR WEAPONS SYSTEMS.
 * -----------------------------------------------------------------------------
 */

#include "PIOPeephole.h"

#include <format>
#include <optional>

std::string PIOAsmLine::to_string() const {
  std::string s;
  switch (type) {
    case INSTRUCTION:
      s = "    " + mnemonic;
      if (!op1.empty()) {
        s += " " + op1;
        if (!op2.empty()) {
          s += ", " + op2;
          if (!op3.empty()) {
            s += ", " + op3;
          }
        }
      }
      if (delay > 0) {
        s += std::format(" [{}]", delay);
      }
      return s;
    case LABEL:
      return std::format("{}:", label);
    case COMMENT:
      return std::format("; {}", content);
    case RAW:
      return content;
    case EMPTY:
      return "";
  }
  return "";
}

std::vector<PIOAsmLine> PIOPeephole::optimize(
    const std::vector<PIOAsmLine> &lines) {
  std::vector<PIOAsmLine> result;
  std::optional<std::string> x_val;
  std::optional<std::string> y_val;

  for (const auto &line : lines) {
    if (line.type == PIOAsmLine::INSTRUCTION) {
      // Remove redundant MOV
      if (line.mnemonic == "MOV" && line.op1 == line.op2) {
        continue;
      }

      // Simple state tracking for SET
      if (line.mnemonic == "SET") {
        if (line.op1 == "X") {
          if (x_val && *x_val == line.op2) continue;
          x_val = line.op2;
        } else if (line.op1 == "Y") {
          if (y_val && *y_val == line.op2) continue;
          y_val = line.op2;
        }
      } else if (line.mnemonic == "MOV") {
        if (line.op1 == "X") {
          if (line.op2 == "Y" && x_val && y_val && *x_val == *y_val) continue;
          x_val.reset();  // Unknown now
        } else if (line.op1 == "Y") {
          if (line.op2 == "X" && x_val && y_val && *x_val == *y_val) continue;
          y_val.reset();
        }
      } else {
        // Any other instruction might change state
        x_val.reset();
        y_val.reset();
      }
    } else if (line.type == PIOAsmLine::LABEL) {
      x_val.reset();
      y_val.reset();
    }
    result.push_back(line);
  }

  return result;
}

File: src/compiler/Backend/Targets/PIO/PIOPeephole.cs
/*
 * -----------------------------------------------------------------------------
 * PyMCU Compiler (pymcuc)
 * Copyright (C) 2026 Ivan Montiel Cardona and the PyMCU Project Authors
 *
 * SPDX-License-Identifier: MIT
 *
 * -----------------------------------------------------------------------------
 * SAFETY WARNING / HIGH RISK ACTIVITIES:
 * THE SOFTWARE IS NOT DESIGNED, MANUFACTURED, OR INTENDED FOR USE IN HAZARDOUS
 * ENVIRONMENTS REQUIRING FAIL-SAFE PERFORMANCE, SUCH AS IN THE OPERATION OF
 * NUCLEAR FACILITIES, AIRCRAFT NAVIGATION OR COMMUNICATION SYSTEMS, AIR
 * TRAFFIC CONTROL, DIRECT LIFE SUPPORT MACHINES, OR WEAPONS SYSTEMS.
 * -----------------------------------------------------------------------------
 */

namespace PyMCU.Backend.Targets.PIO;

public class PIOAsmLine
{
    public enum LineType
    {
        Instruction,
        Label,
        Comment,
        Raw,
        Empty
    }

    public LineType Type;
    public string LabelText = "";
    public string Mnemonic = "";
    public string Op1 = "";
    public string Op2 = "";
    public string Op3 = "";
    public string Content = "";
    public int Delay = 0;

    public static PIOAsmLine MakeInstruction(string m, string o1 = "", string o2 = "", string o3 = "")
        => new() { Type = LineType.Instruction, Mnemonic = m, Op1 = o1, Op2 = o2, Op3 = o3 };

    public static PIOAsmLine MakeLabel(string l)
        => new() { Type = LineType.Label, LabelText = l };

    public static PIOAsmLine MakeComment(string c)
        => new() { Type = LineType.Comment, Content = c };

    public static PIOAsmLine MakeRaw(string r)
        => new() { Type = LineType.Raw, Content = r };

    public static PIOAsmLine MakeEmpty()
        => new() { Type = LineType.Empty };

    public override string ToString()
    {
        switch (Type)
        {
            case LineType.Instruction:
            {
                string s = "    " + Mnemonic;
                if (!string.IsNullOrEmpty(Op1))
                {
                    s += " " + Op1;
                    if (!string.IsNullOrEmpty(Op2))
                    {
                        s += ", " + Op2;
                        if (!string.IsNullOrEmpty(Op3))
                            s += ", " + Op3;
                    }
                }

                if (Delay > 0) s += $" [{Delay}]";
                return s;
            }
            case LineType.Label: return $"{LabelText}:";
            case LineType.Comment: return $"; {Content}";
            case LineType.Raw: return Content;
            default: return "";
        }
    }
}

public static class PIOPeephole
{
    public static List<PIOAsmLine> Optimize(List<PIOAsmLine> lines)
    {
        var result = new List<PIOAsmLine>();
        string? xVal = null;
        string? yVal = null;

        foreach (var line in lines)
        {
            if (line.Type == PIOAsmLine.LineType.Instruction)
            {
                // Remove redundant MOV
                if (line.Mnemonic == "MOV" && line.Op1 == line.Op2) continue;

                if (line.Mnemonic == "SET")
                {
                    if (line.Op1 == "X")
                    {
                        if (xVal == line.Op2) continue;
                        xVal = line.Op2;
                    }
                    else if (line.Op1 == "Y")
                    {
                        if (yVal == line.Op2) continue;
                        yVal = line.Op2;
                    }
                }
                else if (line.Mnemonic == "MOV")
                {
                    if (line.Op1 == "X")
                    {
                        if (line.Op2 == "Y" && xVal != null && yVal != null && xVal == yVal) continue;
                        xVal = null;
                    }
                    else if (line.Op1 == "Y")
                    {
                        if (line.Op2 == "X" && xVal != null && yVal != null && xVal == yVal) continue;
                        yVal = null;
                    }
                }
                else
                {
                    xVal = null;
                    yVal = null;
                }
            }
            else if (line.Type == PIOAsmLine.LineType.Label)
            {
                xVal = null;
                yVal = null;
            }

            result.Add(line);
        }

        return result;
    }
}